### PR TITLE
Reworked version of backend with Yoda.

### PR DESCRIPTION
**Pull request by mzhurovich**

### CODE REVIEW

File: schema.h [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+enum class UID : uint64_t { INVALID = 0u };
+enum class CID : uint64_t { INVALID = 0u };
+enum class ANSWER : int { UNSEEN = 0, CTFO = 1, TFU = 2, TIFB = -1 };
+
+struct User : yoda::Padawan {
+  UID uid;
+  uint64_t score = 0u;
+
+  UID key() const { return uid; }
```
Comment (dkorolev) [inline schema.h]: `const` ?

File: schema.h [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+  void serialize(A& ar) {
+    Padawan::serialize(ar);
+    ar(CEREAL_NVP(score));
+  }
+};
+
+struct UIDTokenPair : yoda::Padawan {
+  UID uid = UID::INVALID;
+  std::string token = "";
+  bool valid = true;
```
Comment (dkorolev) [inline schema.h]: Why `true` by default, here and in constructor?

File: schema.h [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+  UIDTokenPair() = default;
+  UIDTokenPair(const UIDTokenPair&) = default;
+  UIDTokenPair(UID uid, const std::string& token, bool valid = true) : uid(uid), token(token), valid(valid) {}
+
+  UID row() const { return uid; }
+  void set_row(UID value) { uid = value; }
+  const std::string& col() const { return token; }
+  void set_col(std::string value) { token = value; }
```
Comment (dkorolev) [inline schema.h]: `const std::string&` as param.

File: schema.h [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+  template <typename A>
+  void serialize(A& ar) {
+    Padawan::serialize(ar);
+    ar(CEREAL_NVP(uid), CEREAL_NVP(token), CEREAL_NVP(valid));
+  }
+};
+
+struct DeviceIdUIDPair : yoda::Padawan {
+  std::string device_id = "";
+  UID uid;
```
Comment (dkorolev) [inline schema.h]: Y U NO PUT `INVALID`?

File: server.cc [hunk 5]
```diff
@@ -1,6 +1,8 @@
+/*******************************************************************************
 The MIT License (MIT)
 
-Copyright (c) 2015 KnowSheet
+Copyright (c) 2015 Dmitry "Dima" Korolev <dmitry.korolev@gmail.com>
```
Comment (dkorolev) [inline server.cc]: Flip author lines? :)

File: server.h [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        random_uint64_(std::bind(std::uniform_int_distribution<uint64_t>(0, -1ll), rng_)),
+        random_0_1_picker_(std::bind(std::uniform_real_distribution<double>(0.0, 1.0), rng_)),
+        random_10_99_picker_(std::bind(std::uniform_int_distribution<int>(10, 99), rng_)),
+        storage_("CTFO storage"),
+        cards_(Split<ByLines>(bricks::FileSystem::ReadFileAsString("cards.txt"))) {
+    assert(cards_.size() >= 2u);
+    assert(cards_[0] == "ZERO_INDEX_SHOULD_BE_KEPT_UNUSED");
+
+    storage_.Transaction([this](StorageAPI::T_DATA data) {
+      for (const auto text : cards_) {
```
Comment (dkorolev) [inline server.h]: `const auto&` ?

File: server.h [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+              MatrixEntry<DeviceIdUIDPair>,
+              Dictionary<Card>,
+              MatrixEntry<Answer>> StorageAPI;
+  StorageAPI storage_;
+  std::vector<std::string> cards_;
+
+  const std::map<std::string, ANSWER> valid_answers_ = {
+      {"ctfo", ANSWER::CTFO}, {"tfu", ANSWER::TFU}, {"tifb", ANSWER::TIFB}};
+
+  UID RandomUID() { return static_cast<UID>(random_uint64_() % 1000000000000000000 + 1000000000000000000); }
```
Comment (dkorolev) [inline server.h]: Why not use `random_uid_picker_` and `random_cid_picker_` with proper ranges?

File: server.h [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+        cards_(Split<ByLines>(bricks::FileSystem::ReadFileAsString("cards.txt"))) {
+    assert(cards_.size() >= 2u);
+    assert(cards_[0] == "ZERO_INDEX_SHOULD_BE_KEPT_UNUSED");
+
+    storage_.Transaction([this](StorageAPI::T_DATA data) {
+      for (const auto text : cards_) {
+        CID cid;
+        do {
+          cid = RandomCID();
+        } while (data.Get(cid));
```
Comment (dkorolev) [inline server.h]: while `data.Has(cid)`?

File: server.h [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+            } else {
+              UID uid = UID::INVALID;
+              User user;
+              ResponseUserEntry user_entry;
+
+              // Searching for users with provided device ID.
+              storage_.Transaction([&](StorageAPI::T_DATA data) {
+                                     const auto accessor = MatrixEntry<DeviceIdUIDPair>::Accessor(data);
+                                     if (accessor.Rows().has(device_id)) {
+                                       assert(accessor[device_id].size() == 1);
```
Comment (dkorolev) [inline server.h]: Pls add a comment about why is this bad. Because it really is.

File: server.h [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+                                     }
+                                   }).Go();
+
+              if (uid != UID::INVALID) {  // Existing user.
+                // Invalidating all old tokens.
+                storage_.Transaction([&](StorageAPI::T_DATA data) {
+                                       auto mutator = MatrixEntry<UIDTokenPair>::Mutator(data);
+                                       for (const auto& uid_token : mutator[uid]) {
+                                         UIDTokenPair invalid_token(uid_token.uid, uid_token.token, false);
+                                         mutator.Add(invalid_token);
```
Comment (dkorolev) [inline server.h]: You can save on a variable here.

File: server.h [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+                                     }).Go();
+                user_entry.score = user.score;
+              } else {  // New user.
+                uid = RandomUID();
+                user.uid = uid;
+                storage_.Add(user).Go();
+              }
+
+              CopyUserInfoToResponseEntry(user, user_entry);
+              // Generating a new token.
```
Comment (dkorolev) [inline server.h]: Nit: `Generate`. Just a nit.

File: server.h [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+                ResponseUserEntry user_entry;
+                CopyUserInfoToResponseEntry(user, user_entry);
+                user_entry.token = token;
+                RespondWithFeed(user_entry, FromString<size_t>(r.url.query.get("feed_count", "20")),
+                                std::move(r));
+              }
+            }
+          }
+        });
+  }  // CTFOServer(...)
```
Comment (dkorolev) [inline server.h]: Don't think this comment is useful here.

File: server.h [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+
+  void RespondWithFeed(ResponseUserEntry user_entry, size_t max_count, Request r) {
+    storage_.Transaction(
+        [&, user_entry, max_count](StorageAPI::T_DATA data) {
```
Comment (dkorolev) [inline server.h]: I remember we chat about it, yet, once again: what does the `[&]` capture here and can it be avoided?

File: test.cc [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+CEREAL_REGISTER_TYPE(UIDTokenPair);
+CEREAL_REGISTER_TYPE(DeviceIdUIDPair);
+CEREAL_REGISTER_TYPE(Card);
+CEREAL_REGISTER_TYPE(Answer);
+
+DEFINE_int32(port, 8383, "Port to spawn CTFO backend server on.");
+DEFINE_int32(rand_seed, 42, "The answer to the question of life, universe and everything.");
+
+TEST(CTFO, SmokeTest) {
+  const int port = FLAGS_port;
```
Comment (dkorolev) [inline test.cc]: I'm positive you don't need these two variables :)

File: test.cc [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+  const int seed = FLAGS_rand_seed;
+  CTFOServer server(port, seed);
+  bricks::time::SetNow(static_cast<bricks::time::EPOCH_MILLISECONDS>(123));
+
+  const std::string device_id_str = "A_BUNCH_OF_NUMBERS";
```
Comment (dkorolev) [inline test.cc]: Of digits, bro :sunglasses:

File: test.cc [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+  bricks::time::SetNow(static_cast<bricks::time::EPOCH_MILLISECONDS>(234));
+
+  const auto feed_response = HTTP(
+      GET(Printf("http://localhost:%d/feed?uid=%s&token=%s&feed_count=40", port, golden_uid, golden_token)));
+  EXPECT_EQ(200, static_cast<int>(feed_response.code));
+  feed = ParseJSON<ResponseFeed>(feed_response.body);
+  EXPECT_EQ(234u, feed.ts);
+  EXPECT_EQ(golden_uid_str, feed.user.uid);
+  EXPECT_EQ(golden_token_str, feed.user.token);
+  EXPECT_EQ(40u, feed.cards.size());
```
Comment (dkorolev) [inline test.cc:86]: I'd check a couple extra things here. At least that the cards have non-empty texts, that are all different, as well as their ID-s.

**Comment by dkorolev** [comment]
**LIKE++**!

I have a surprisingly small number of comments given the amount of new code, and the fact that the 1st time it is Yoda production sees.

Let's sync up tomorrow.

Thanks!
Dima

File: schema.h [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+#define CTFO_SCHEMA_H
+
+#include <vector>
+#include <string>
+
+#include "../Current/Bricks/cerealize/cerealize.h"
+#include "../Current/Sherlock/yoda/yoda.h"
+
+// Data structures for internal storage.
+
```
Comment (dkorolev) [inline schema.h]: This vertical whitespace is redundant.

File: schema.h [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+  template <typename A>
+  void serialize(A& ar) {
+    Padawan::serialize(ar);
+    ar(CEREAL_NVP(uid), CEREAL_NVP(cid), CEREAL_NVP(answer));
+  }
+};
+
+// Data structures for generating RESTful response.
+
```
Comment (dkorolev) [inline schema.h]: This vertical whitespace is redundant.

File: server.h (hunk 13 above)
Comment (mzhurovich) [inline server.h]: It captures `this` and provides access to random number generators.

Comment (dkorolev) [inline server.h]: I see -- would `[this, ... ]`, or `auto& rng = rng_; [this, &rng, ...]` work?

_/me is still against `[&]` and `[=]` captures._

**Comment by mzhurovich** [comment]
@dkorolev Thanks! PTAL.

**Comment by dkorolev** [comment]
@mzhurovich LGTM++!

Two bits:
- https://github.com/mzhurovich/CTFO_BE/commit/bc45df1863ba7c90cb98fea5977ce7ef3e5b121b#commitcomment-11554601
- https://github.com/mzhurovich/CTFO_BE/commit/bc45df1863ba7c90cb98fea5977ce7ef3e5b121b#commitcomment-11554596

Thanks,
Dima

**Comment by mzhurovich** [comment]
@dkorolev Fixed!

**Comment by dkorolev** [comment]
Yay! Merging.

**Pull request merged**